{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.extract gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/lsg/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    DDX11L2  MIR1302-2HG  FAM138A  ENSG00000290826  OR4F5  \\\n",
      "AAACCTGAGATGTCGG-1      0.0          0.0      0.0              0.0    0.0   \n",
      "AAACCTGAGCCAGTTT-1      0.0          0.0      0.0              0.0    0.0   \n",
      "AAACCTGCATGGAATA-1      0.0          0.0      0.0              0.0    0.0   \n",
      "AAACCTGCATTGGGCC-1      0.0          0.0      0.0              0.0    0.0   \n",
      "AAACCTGGTGCTAGCC-1      0.0          0.0      0.0              0.0    0.0   \n",
      "\n",
      "                    ENSG00000238009  ENSG00000239945  ENSG00000239906  \\\n",
      "AAACCTGAGATGTCGG-1              0.0              0.0              0.0   \n",
      "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0   \n",
      "AAACCTGCATGGAATA-1              0.0              0.0              0.0   \n",
      "AAACCTGCATTGGGCC-1              0.0              0.0              0.0   \n",
      "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0   \n",
      "\n",
      "                    ENSG00000241860  ENSG00000241599  ...  ENSG00000275249  \\\n",
      "AAACCTGAGATGTCGG-1              0.0              0.0  ...              0.0   \n",
      "AAACCTGAGCCAGTTT-1              0.0              0.0  ...              0.0   \n",
      "AAACCTGCATGGAATA-1              0.0              0.0  ...              0.0   \n",
      "AAACCTGCATTGGGCC-1              0.0              0.0  ...              0.0   \n",
      "AAACCTGGTGCTAGCC-1              0.0              0.0  ...              0.0   \n",
      "\n",
      "                    ENSG00000274792  ENSG00000274175  ENSG00000275869  \\\n",
      "AAACCTGAGATGTCGG-1              0.0              0.0              0.0   \n",
      "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0   \n",
      "AAACCTGCATGGAATA-1              0.0              0.0              0.0   \n",
      "AAACCTGCATTGGGCC-1              0.0              0.0              0.0   \n",
      "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0   \n",
      "\n",
      "                    ENSG00000273554  ENSG00000277836  ENSG00000278633  \\\n",
      "AAACCTGAGATGTCGG-1              0.0              0.0              0.0   \n",
      "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0   \n",
      "AAACCTGCATGGAATA-1              0.0              0.0              0.0   \n",
      "AAACCTGCATTGGGCC-1              0.0              0.0              0.0   \n",
      "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0   \n",
      "\n",
      "                    ENSG00000276017  ENSG00000278817  ENSG00000277196  \n",
      "AAACCTGAGATGTCGG-1              0.0              0.0              0.0  \n",
      "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0  \n",
      "AAACCTGCATGGAATA-1              0.0              0.0              0.0  \n",
      "AAACCTGCATTGGGCC-1              0.0              0.0              0.0  \n",
      "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0  \n",
      "\n",
      "[5 rows x 38606 columns]\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "# Assuming 'adata' is your AnnData object\n",
    "# Extract the normalized expression matrix\n",
    "adata = sc.read_h5ad(\"/home/lsg/Data/glioblastoma/output/new/h5ad/SC-bbknn.h5ad\")\n",
    "normalized_data = pd.DataFrame(adata.X.toarray(), \n",
    "                               index=adata.obs_names, \n",
    "                               columns=adata.var_names)\n",
    "\n",
    "# View or save the data\n",
    "print(normalized_data.head())  # Display the first few rows\n",
    "normalized_data.to_csv(\"gene_expression.csv\")  # Save as CSV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.pyWGCNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92mSaving data to be True, checking requirements ...\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DDX11L2</th>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <th>FAM138A</th>\n",
       "      <th>ENSG00000290826</th>\n",
       "      <th>OR4F5</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <th>ENSG00000239906</th>\n",
       "      <th>ENSG00000241860</th>\n",
       "      <th>ENSG00000241599</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000275249</th>\n",
       "      <th>ENSG00000274792</th>\n",
       "      <th>ENSG00000274175</th>\n",
       "      <th>ENSG00000275869</th>\n",
       "      <th>ENSG00000273554</th>\n",
       "      <th>ENSG00000277836</th>\n",
       "      <th>ENSG00000278633</th>\n",
       "      <th>ENSG00000276017</th>\n",
       "      <th>ENSG00000278817</th>\n",
       "      <th>ENSG00000277196</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGATGTCGG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCCAGTTT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGGAATA-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATTGGGCC-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGCTAGCC-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 38606 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    DDX11L2  MIR1302-2HG  FAM138A  ENSG00000290826  OR4F5  \\\n",
       "AAACCTGAGATGTCGG-1      0.0          0.0      0.0              0.0    0.0   \n",
       "AAACCTGAGCCAGTTT-1      0.0          0.0      0.0              0.0    0.0   \n",
       "AAACCTGCATGGAATA-1      0.0          0.0      0.0              0.0    0.0   \n",
       "AAACCTGCATTGGGCC-1      0.0          0.0      0.0              0.0    0.0   \n",
       "AAACCTGGTGCTAGCC-1      0.0          0.0      0.0              0.0    0.0   \n",
       "\n",
       "                    ENSG00000238009  ENSG00000239945  ENSG00000239906  \\\n",
       "AAACCTGAGATGTCGG-1              0.0              0.0              0.0   \n",
       "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0   \n",
       "AAACCTGCATGGAATA-1              0.0              0.0              0.0   \n",
       "AAACCTGCATTGGGCC-1              0.0              0.0              0.0   \n",
       "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0   \n",
       "\n",
       "                    ENSG00000241860  ENSG00000241599  ...  ENSG00000275249  \\\n",
       "AAACCTGAGATGTCGG-1              0.0              0.0  ...              0.0   \n",
       "AAACCTGAGCCAGTTT-1              0.0              0.0  ...              0.0   \n",
       "AAACCTGCATGGAATA-1              0.0              0.0  ...              0.0   \n",
       "AAACCTGCATTGGGCC-1              0.0              0.0  ...              0.0   \n",
       "AAACCTGGTGCTAGCC-1              0.0              0.0  ...              0.0   \n",
       "\n",
       "                    ENSG00000274792  ENSG00000274175  ENSG00000275869  \\\n",
       "AAACCTGAGATGTCGG-1              0.0              0.0              0.0   \n",
       "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0   \n",
       "AAACCTGCATGGAATA-1              0.0              0.0              0.0   \n",
       "AAACCTGCATTGGGCC-1              0.0              0.0              0.0   \n",
       "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0   \n",
       "\n",
       "                    ENSG00000273554  ENSG00000277836  ENSG00000278633  \\\n",
       "AAACCTGAGATGTCGG-1              0.0              0.0              0.0   \n",
       "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0   \n",
       "AAACCTGCATGGAATA-1              0.0              0.0              0.0   \n",
       "AAACCTGCATTGGGCC-1              0.0              0.0              0.0   \n",
       "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0   \n",
       "\n",
       "                    ENSG00000276017  ENSG00000278817  ENSG00000277196  \n",
       "AAACCTGAGATGTCGG-1              0.0              0.0              0.0  \n",
       "AAACCTGAGCCAGTTT-1              0.0              0.0              0.0  \n",
       "AAACCTGCATGGAATA-1              0.0              0.0              0.0  \n",
       "AAACCTGCATTGGGCC-1              0.0              0.0              0.0  \n",
       "AAACCTGGTGCTAGCC-1              0.0              0.0              0.0  \n",
       "\n",
       "[5 rows x 38606 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import PyWGCNA\n",
    "geneExp = '/home/lsg/Data/glioblastoma/gene_expression.csv'\n",
    "pyWGCNA_5xFAD = PyWGCNA.WGCNA(name='5xFAD', \n",
    "                              species='human', \n",
    "                              geneExpPath=geneExp, \n",
    "                              outputPath='',\n",
    "                              save=True)\n",
    "pyWGCNA_5xFAD.geneExpr.to_df().head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[94mPre-processing...\u001b[0m\n",
      "\tDetecting genes and samples with too many missing values...\n"
     ]
    },
    {
     "ename": "RecursionError",
     "evalue": "maximum recursion depth exceeded in comparison",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRecursionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mpyWGCNA_5xFAD\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpreprocess\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/PyWGCNA/wgcna.py:243\u001b[0m, in \u001b[0;36mWGCNA.preprocess\u001b[0;34m(self, show)\u001b[0m\n\u001b[1;32m    241\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m show:\n\u001b[1;32m    242\u001b[0m     plt\u001b[38;5;241m.\u001b[39mfigure(figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;28mmax\u001b[39m(\u001b[38;5;241m25\u001b[39m, \u001b[38;5;28mround\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdatExpr\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m/\u001b[39m \u001b[38;5;241m20\u001b[39m)), \u001b[38;5;241m10\u001b[39m), facecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwhite\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 243\u001b[0m     \u001b[43mdendrogram\u001b[49m\u001b[43m(\u001b[49m\u001b[43msampleTree\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolor_threshold\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcut\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlabels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdatExpr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_df\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mleaf_rotation\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m90\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m               \u001b[49m\u001b[43mleaf_font_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m8\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    245\u001b[0m     plt\u001b[38;5;241m.\u001b[39maxhline(y\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcut, c\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgrey\u001b[39m\u001b[38;5;124m'\u001b[39m, lw\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, linestyle\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdashed\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    246\u001b[0m     plt\u001b[38;5;241m.\u001b[39mtitle(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSample clustering to detect outliers\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3332\u001b[0m, in \u001b[0;36mdendrogram\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, no_plot, no_labels, leaf_font_size, leaf_rotation, leaf_label_func, show_contracted, link_color_func, ax, above_threshold_color)\u001b[0m\n\u001b[1;32m   3330\u001b[0m if get_leaves:\n\u001b[1;32m   3331\u001b[0m     lvs = []\n\u001b[0;32m-> 3332\u001b[0m else:\n\u001b[1;32m   3333\u001b[0m     lvs = None\n\u001b[1;32m   3335\u001b[0m icoord_list = []\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3609\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3606\u001b[0m         ua = aa\n\u001b[1;32m   3607\u001b[0m         ub = ab\n\u001b[1;32m   3608\u001b[0m elif distance_sort == 'descending':\n\u001b[0;32m-> 3609\u001b[0m     # If a has a distance less than or equal to b, it and its\n\u001b[1;32m   3610\u001b[0m     # descendents should be drawn to the left. Otherwise, to\n\u001b[1;32m   3611\u001b[0m     # the right.\n\u001b[1;32m   3612\u001b[0m     if da > db:\n\u001b[1;32m   3613\u001b[0m         ua = aa\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3642\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3620\u001b[0m     ub \u001b[38;5;241m=\u001b[39m ab\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;66;03m# Updated iv variable and the amount of space used.\u001b[39;00m\n\u001b[1;32m   3623\u001b[0m (uiva, uwa, uah, uamd) \u001b[38;5;241m=\u001b[39m \\\n\u001b[1;32m   3624\u001b[0m     _dendrogram_calculate_info(\n\u001b[1;32m   3625\u001b[0m         Z\u001b[38;5;241m=\u001b[39mZ, p\u001b[38;5;241m=\u001b[39mp,\n\u001b[1;32m   3626\u001b[0m         truncate_mode\u001b[38;5;241m=\u001b[39mtruncate_mode,\n\u001b[1;32m   3627\u001b[0m         color_threshold\u001b[38;5;241m=\u001b[39mcolor_threshold,\n\u001b[1;32m   3628\u001b[0m         get_leaves\u001b[38;5;241m=\u001b[39mget_leaves,\n\u001b[1;32m   3629\u001b[0m         orientation\u001b[38;5;241m=\u001b[39morientation,\n\u001b[1;32m   3630\u001b[0m         labels\u001b[38;5;241m=\u001b[39mlabels,\n\u001b[1;32m   3631\u001b[0m         count_sort\u001b[38;5;241m=\u001b[39mcount_sort,\n\u001b[1;32m   3632\u001b[0m         distance_sort\u001b[38;5;241m=\u001b[39mdistance_sort,\n\u001b[1;32m   3633\u001b[0m         show_leaf_counts\u001b[38;5;241m=\u001b[39mshow_leaf_counts,\n\u001b[1;32m   3634\u001b[0m         i\u001b[38;5;241m=\u001b[39mua, iv\u001b[38;5;241m=\u001b[39miv, ivl\u001b[38;5;241m=\u001b[39mivl, n\u001b[38;5;241m=\u001b[39mn,\n\u001b[1;32m   3635\u001b[0m         icoord_list\u001b[38;5;241m=\u001b[39micoord_list,\n\u001b[1;32m   3636\u001b[0m         dcoord_list\u001b[38;5;241m=\u001b[39mdcoord_list, lvs\u001b[38;5;241m=\u001b[39mlvs,\n\u001b[1;32m   3637\u001b[0m         current_color\u001b[38;5;241m=\u001b[39mcurrent_color,\n\u001b[1;32m   3638\u001b[0m         color_list\u001b[38;5;241m=\u001b[39mcolor_list,\n\u001b[1;32m   3639\u001b[0m         currently_below_threshold\u001b[38;5;241m=\u001b[39mcurrently_below_threshold,\n\u001b[1;32m   3640\u001b[0m         leaf_label_func\u001b[38;5;241m=\u001b[39mleaf_label_func,\n\u001b[1;32m   3641\u001b[0m         level\u001b[38;5;241m=\u001b[39mlevel \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m, contraction_marks\u001b[38;5;241m=\u001b[39mcontraction_marks,\n\u001b[0;32m-> 3642\u001b[0m         link_color_func\u001b[38;5;241m=\u001b[39mlink_color_func,\n\u001b[1;32m   3643\u001b[0m         above_threshold_color\u001b[38;5;241m=\u001b[39mabove_threshold_color)\n\u001b[1;32m   3645\u001b[0m h \u001b[38;5;241m=\u001b[39m Z[i \u001b[38;5;241m-\u001b[39m n, \u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m   3646\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m h \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m color_threshold \u001b[38;5;129;01mor\u001b[39;00m color_threshold \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3642\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3620\u001b[0m     ub \u001b[38;5;241m=\u001b[39m ab\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;66;03m# Updated iv variable and the amount of space used.\u001b[39;00m\n\u001b[1;32m   3623\u001b[0m (uiva, uwa, uah, uamd) \u001b[38;5;241m=\u001b[39m \\\n\u001b[1;32m   3624\u001b[0m     _dendrogram_calculate_info(\n\u001b[1;32m   3625\u001b[0m         Z\u001b[38;5;241m=\u001b[39mZ, p\u001b[38;5;241m=\u001b[39mp,\n\u001b[1;32m   3626\u001b[0m         truncate_mode\u001b[38;5;241m=\u001b[39mtruncate_mode,\n\u001b[1;32m   3627\u001b[0m         color_threshold\u001b[38;5;241m=\u001b[39mcolor_threshold,\n\u001b[1;32m   3628\u001b[0m         get_leaves\u001b[38;5;241m=\u001b[39mget_leaves,\n\u001b[1;32m   3629\u001b[0m         orientation\u001b[38;5;241m=\u001b[39morientation,\n\u001b[1;32m   3630\u001b[0m         labels\u001b[38;5;241m=\u001b[39mlabels,\n\u001b[1;32m   3631\u001b[0m         count_sort\u001b[38;5;241m=\u001b[39mcount_sort,\n\u001b[1;32m   3632\u001b[0m         distance_sort\u001b[38;5;241m=\u001b[39mdistance_sort,\n\u001b[1;32m   3633\u001b[0m         show_leaf_counts\u001b[38;5;241m=\u001b[39mshow_leaf_counts,\n\u001b[1;32m   3634\u001b[0m         i\u001b[38;5;241m=\u001b[39mua, iv\u001b[38;5;241m=\u001b[39miv, ivl\u001b[38;5;241m=\u001b[39mivl, n\u001b[38;5;241m=\u001b[39mn,\n\u001b[1;32m   3635\u001b[0m         icoord_list\u001b[38;5;241m=\u001b[39micoord_list,\n\u001b[1;32m   3636\u001b[0m         dcoord_list\u001b[38;5;241m=\u001b[39mdcoord_list, lvs\u001b[38;5;241m=\u001b[39mlvs,\n\u001b[1;32m   3637\u001b[0m         current_color\u001b[38;5;241m=\u001b[39mcurrent_color,\n\u001b[1;32m   3638\u001b[0m         color_list\u001b[38;5;241m=\u001b[39mcolor_list,\n\u001b[1;32m   3639\u001b[0m         currently_below_threshold\u001b[38;5;241m=\u001b[39mcurrently_below_threshold,\n\u001b[1;32m   3640\u001b[0m         leaf_label_func\u001b[38;5;241m=\u001b[39mleaf_label_func,\n\u001b[1;32m   3641\u001b[0m         level\u001b[38;5;241m=\u001b[39mlevel \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m, contraction_marks\u001b[38;5;241m=\u001b[39mcontraction_marks,\n\u001b[0;32m-> 3642\u001b[0m         link_color_func\u001b[38;5;241m=\u001b[39mlink_color_func,\n\u001b[1;32m   3643\u001b[0m         above_threshold_color\u001b[38;5;241m=\u001b[39mabove_threshold_color)\n\u001b[1;32m   3645\u001b[0m h \u001b[38;5;241m=\u001b[39m Z[i \u001b[38;5;241m-\u001b[39m n, \u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m   3646\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m h \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m color_threshold \u001b[38;5;129;01mor\u001b[39;00m color_threshold \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "    \u001b[0;31m[... skipping similar frames: _dendrogram_calculate_info at line 3642 (1790 times)]\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3609\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3606\u001b[0m         ua = aa\n\u001b[1;32m   3607\u001b[0m         ub = ab\n\u001b[1;32m   3608\u001b[0m elif distance_sort == 'descending':\n\u001b[0;32m-> 3609\u001b[0m     # If a has a distance less than or equal to b, it and its\n\u001b[1;32m   3610\u001b[0m     # descendents should be drawn to the left. Otherwise, to\n\u001b[1;32m   3611\u001b[0m     # the right.\n\u001b[1;32m   3612\u001b[0m     if da > db:\n\u001b[1;32m   3613\u001b[0m         ua = aa\n",
      "    \u001b[0;31m[... skipping similar frames: _dendrogram_calculate_info at line 3642 (1169 times), _dendrogram_calculate_info at line 3609 (1 times)]\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3609\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3606\u001b[0m         ua = aa\n\u001b[1;32m   3607\u001b[0m         ub = ab\n\u001b[1;32m   3608\u001b[0m elif distance_sort == 'descending':\n\u001b[0;32m-> 3609\u001b[0m     # If a has a distance less than or equal to b, it and its\n\u001b[1;32m   3610\u001b[0m     # descendents should be drawn to the left. Otherwise, to\n\u001b[1;32m   3611\u001b[0m     # the right.\n\u001b[1;32m   3612\u001b[0m     if da > db:\n\u001b[1;32m   3613\u001b[0m         ua = aa\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3642\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3620\u001b[0m     ub \u001b[38;5;241m=\u001b[39m ab\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;66;03m# Updated iv variable and the amount of space used.\u001b[39;00m\n\u001b[1;32m   3623\u001b[0m (uiva, uwa, uah, uamd) \u001b[38;5;241m=\u001b[39m \\\n\u001b[1;32m   3624\u001b[0m     _dendrogram_calculate_info(\n\u001b[1;32m   3625\u001b[0m         Z\u001b[38;5;241m=\u001b[39mZ, p\u001b[38;5;241m=\u001b[39mp,\n\u001b[1;32m   3626\u001b[0m         truncate_mode\u001b[38;5;241m=\u001b[39mtruncate_mode,\n\u001b[1;32m   3627\u001b[0m         color_threshold\u001b[38;5;241m=\u001b[39mcolor_threshold,\n\u001b[1;32m   3628\u001b[0m         get_leaves\u001b[38;5;241m=\u001b[39mget_leaves,\n\u001b[1;32m   3629\u001b[0m         orientation\u001b[38;5;241m=\u001b[39morientation,\n\u001b[1;32m   3630\u001b[0m         labels\u001b[38;5;241m=\u001b[39mlabels,\n\u001b[1;32m   3631\u001b[0m         count_sort\u001b[38;5;241m=\u001b[39mcount_sort,\n\u001b[1;32m   3632\u001b[0m         distance_sort\u001b[38;5;241m=\u001b[39mdistance_sort,\n\u001b[1;32m   3633\u001b[0m         show_leaf_counts\u001b[38;5;241m=\u001b[39mshow_leaf_counts,\n\u001b[1;32m   3634\u001b[0m         i\u001b[38;5;241m=\u001b[39mua, iv\u001b[38;5;241m=\u001b[39miv, ivl\u001b[38;5;241m=\u001b[39mivl, n\u001b[38;5;241m=\u001b[39mn,\n\u001b[1;32m   3635\u001b[0m         icoord_list\u001b[38;5;241m=\u001b[39micoord_list,\n\u001b[1;32m   3636\u001b[0m         dcoord_list\u001b[38;5;241m=\u001b[39mdcoord_list, lvs\u001b[38;5;241m=\u001b[39mlvs,\n\u001b[1;32m   3637\u001b[0m         current_color\u001b[38;5;241m=\u001b[39mcurrent_color,\n\u001b[1;32m   3638\u001b[0m         color_list\u001b[38;5;241m=\u001b[39mcolor_list,\n\u001b[1;32m   3639\u001b[0m         currently_below_threshold\u001b[38;5;241m=\u001b[39mcurrently_below_threshold,\n\u001b[1;32m   3640\u001b[0m         leaf_label_func\u001b[38;5;241m=\u001b[39mleaf_label_func,\n\u001b[1;32m   3641\u001b[0m         level\u001b[38;5;241m=\u001b[39mlevel \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m, contraction_marks\u001b[38;5;241m=\u001b[39mcontraction_marks,\n\u001b[0;32m-> 3642\u001b[0m         link_color_func\u001b[38;5;241m=\u001b[39mlink_color_func,\n\u001b[1;32m   3643\u001b[0m         above_threshold_color\u001b[38;5;241m=\u001b[39mabove_threshold_color)\n\u001b[1;32m   3645\u001b[0m h \u001b[38;5;241m=\u001b[39m Z[i \u001b[38;5;241m-\u001b[39m n, \u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m   3646\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m h \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m color_threshold \u001b[38;5;129;01mor\u001b[39;00m color_threshold \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:3546\u001b[0m, in \u001b[0;36m_dendrogram_calculate_info\u001b[0;34m(Z, p, truncate_mode, color_threshold, get_leaves, orientation, labels, count_sort, distance_sort, show_leaf_counts, i, iv, ivl, n, icoord_list, dcoord_list, lvs, mhr, current_color, color_list, currently_below_threshold, leaf_label_func, level, contraction_marks, link_color_func, above_threshold_color)\u001b[0m\n\u001b[1;32m   3544\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m (iv \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m5.0\u001b[39m, \u001b[38;5;241m10.0\u001b[39m, \u001b[38;5;241m0.0\u001b[39m, d)\n\u001b[1;32m   3545\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m i \u001b[38;5;241m<\u001b[39m n:\n\u001b[0;32m-> 3546\u001b[0m         _append_singleton_leaf_node(Z, p, n, level, lvs, ivl,\n\u001b[1;32m   3547\u001b[0m                                     leaf_label_func, i, labels)\n\u001b[1;32m   3548\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m (iv \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m5.0\u001b[39m, \u001b[38;5;241m10.0\u001b[39m, \u001b[38;5;241m0.0\u001b[39m, \u001b[38;5;241m0.0\u001b[39m)\n\u001b[1;32m   3550\u001b[0m \u001b[38;5;66;03m# Otherwise, only truncate if we have a leaf node.\u001b[39;00m\n\u001b[1;32m   3551\u001b[0m \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m   3552\u001b[0m \u001b[38;5;66;03m# Only place leaves if they correspond to original observations.\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/cluster/hierarchy.py:166\u001b[0m, in \u001b[0;36mint_floor\u001b[0;34m(arr, xp)\u001b[0m\n\u001b[1;32m    162\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_copy_array_if_base_present\u001b[39m(a):\n\u001b[1;32m    163\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    164\u001b[0m \u001b[38;5;124;03m    Copy the array if its base points to a parent array.\u001b[39;00m\n\u001b[1;32m    165\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 166\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m a\u001b[38;5;241m.\u001b[39mbase \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    167\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m a\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m    168\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m np\u001b[38;5;241m.\u001b[39missubsctype(a, np\u001b[38;5;241m.\u001b[39mfloat32):\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/_lib/array_api_compat/common/_aliases.py:315\u001b[0m, in \u001b[0;36m_asarray\u001b[0;34m(obj, dtype, device, copy, namespace, **kwargs)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/envs/scRNA-seq/lib/python3.11/site-packages/scipy/_lib/array_api_compat/common/_helpers.py:296\u001b[0m, in \u001b[0;36m_check_device\u001b[0;34m(xp, device)\u001b[0m\n",
      "\u001b[0;31mRecursionError\u001b[0m: maximum recursion depth exceeded in comparison"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 62900x1000 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pyWGCNA_5xFAD.preprocess()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA-seq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
